[DRAFT] Improve efficiency of applying policy map changes

joamaki opened the pull request:
This moves the incremental policy map change applying into the "sync-policy-map" controller instead of performing the applying of the changes directly. `EndpointManager.UpdatePolicyMaps` is refactored to trigger the controller of each endpoint and then to wait for reconciliation to finish. Semantics with regards to when e.g. `UpdatePolicyMaps` returns are unchanged, e.g. a failed reconciliation attempt does not block. Compared to prior a failed incremental policy map update is now retried by the controller.

This will reduce the amount of overhead in processing of ToFQDN policies in relation to a DNS request/response. Prior to this each DNS response would have triggered call to `EndpointManager.UpdatePolicyMaps` which in turn spawned a goroutine per endpoint to apply the changes. With high rate of DNS requests this may lead to lots of endpoint lock contention and context switching overhead due to queueing of calls to `ApplyPolicyMaps` that are unnecessary since a prior call likely had already applied the pending changes.

joamaki committed:
endpoint: Incremental policy map updates via sync-policy-map controller

The previous fix to avoid full dumping of policy map on every policy map change
(867e41a610) missed the likely accidental use of the sync-policy-map controller
for retrying failures in ApplyPolicyMapChanges.

Fix this by using the sync-policy-map controller to apply pending changes to
the endpoint policy map. Use the controller's executed channel for signalling
completion to coalesce multiple requests to apply changes into a single one to reduce overhead.

Set a minimum retry wait time of 100 millis and a maximum of 1 minute. We want to
avoid aggressive retrying when the policy map is full.

Refactor UpdatePolicyMaps to leverage the triggering behaviour and to avoid
spawning a goroutine per endpoint.

Fixes: 867e41a610 ("endpoint: Only perform full synchronization periodically")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/endpoint/bpf.go:
```diff
@@ -1347,6 +1350,11 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 	}
 
+	// Skip full reconciliation if it has been performed recently.
+	if time.Since(e.policyMapLastFullReconciliation) < option.Config.PolicyMapFullReconciliationInterval {
```
can we do this with a trigger and a min-interval instead?

joamaki replied:
Not really. We want to run it on every `PolicyMapFullReconciliationInterval` and we want the visibility that `cilium status --all-controllers` gives us.

bimmlerd replied:
ack, yeah tried to think it through and it doesn't seem to work.

bimmlerd commented on pkg/endpoint/bpf.go:
```diff
@@ -1120,28 +1126,23 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
-func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) error {
+// identities added / deleted) by triggering the sync-policy-map controller.
+// Returns a channel that is closed when the changes are applied. The channel
+// is closed even if errors occurred. The controller will retry the changes.
```
does this mean the channel is closed before the policy is plumbed in case of errors? Or just that it is eventually closed, even in the face of errors?

joamaki replied:
The old semantics returned an error that was logged and then unblocked waiting in `notifyOnDNSMsg`, so that's what I'm keeping here: Close the channel when an attempt has been made to reconcile the policy map changes. 

It's a separate topic whether that should change and we should keep waiting for retry to succeed in `notifyOnDNSMsg`.

bimmlerd commented on pkg/endpoint/bpf.go:
```diff
@@ -1120,28 +1126,23 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
+func (e *Endpoint) ApplyPolicyMapChanges() <-chan struct{} {
 	if err := e.lockAlive(); err != nil {
-		return err
+		return nil
```
1. reading from a nil channel blocks the reader forever; 
2. there's a window between the calls to `GetEndpoints` and `e.ApplyPolicyMapChanges` below

hence this can block the controller until the context expires, which isn't ideal.

joamaki replied:
Damn it, misremembered the semantics. Will fix to return a closed channel.

bimmlerd commented on pkg/endpoint/bpf.go:
```diff
@@ -1406,14 +1427,38 @@ func (e *Endpoint) startSyncPolicyMapController() {
 					return controller.NewExitReason("Endpoint disappeared")
```
if we exit here, do we also need to close the current channel?

joamaki replied:
Good point, it would be good to close it. I'll put it into the `StopFunc` for the controller.

joamaki commented on pkg/endpoint/bpf.go:
```diff
@@ -1406,14 +1427,38 @@ func (e *Endpoint) startSyncPolicyMapController() {
+				// Apply policy changes to L7 proxy
+				proxyWaitGroup := completion.NewWaitGroup(ctx)
+
+				// Ignoring the revertFunc; keep all successful changes even if some fail.
```
@jrajahalme what do you think about the move of the `updateNetworkPolicy` here from `EndpointManager.UpdatePolicyMaps`? And do you think we should return the error here so that the controller will keep retrying this?

jrajahalme replied:
Can you be more specific? `EndpointManager.UpdatePolicyMaps`currently does not call `updateNetworkPolicy`, but it calls `ep.ApplyPolicyMapChanges`.

joamaki replied:
Oops sorry, I meant moving it from `Endpoint.ApplyPolicyMapChanges` to here. Effectively this means we'll call `updateNetworkPolicy` not only on the incremental path, but also during the periodic full reconciliation. So to rephrase the question: is it OK to call `updateNetworkPolicy` every 15 minutes even if there's no changes, or do we need some extra around this to avoid unnecessary calls?

Second question: should the 'sync-policy-map' controller retry failed calls to `updateNetworkPolicy`?

joamaki commented on pkg/endpoint/bpf.go:
```diff
@@ -1406,14 +1427,38 @@ func (e *Endpoint) startSyncPolicyMapController() {
-			RunInterval: option.Config.PolicyMapFullReconciliationInterval,
-			Context:     e.aliveCtx,
+			RunInterval:            option.Config.PolicyMapFullReconciliationInterval,
+			MaxRetryInterval:       time.Minute,
```
Should the retry backoff parameters be configurable? What are reasonable defaults for these?

joamaki commented on pkg/endpoint/bpf.go:
```diff
@@ -1406,14 +1427,38 @@ func (e *Endpoint) startSyncPolicyMapController() {
-				return e.syncPolicyMapWithDump()
+				defer e.swapPolicyMapApplied()
+
+				err := e.syncPolicyMapWithDump(ctx)
```
Any thoughts on the other `syncPolicyMap` methods and how they're used (e.g. regenerateBPF)? Can we fold those into this one as well?

joamaki committed:
channels: Add channels package for common channel-related utilities

DoneChan: the "chan struct{}" pattern for signaling completion is used all over the place
and thus giving it a type alias can be beneficial in type signatures to underline
its usage.

ClosedDoneChan: when a "done channel" is used where nothing needs to be done, a new channel
needs to be created and immediately closed. This pattern seems to repeat a lot. We can instead
create a singleton DoneChan and close it at init time.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/endpoint/bpf.go:
```diff
@@ -1120,25 +1126,25 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
+func (e *Endpoint) ApplyPolicyMapChanges() <-chan struct{} {
 	if err := e.lockAlive(); err != nil {
-		return err
+		ch := make(chan struct{})
```
this then wants to be the `ClosedDoneChan` too

bimmlerd commented on pkg/fqdn/name_manager_test.go:
```diff
@@ -30,11 +29,11 @@ func (ds *FQDNTestSuite) TestNameManagerCIDRGeneration(c *C) {
 					selIPMap[k] = v
 				}
-				return &sync.WaitGroup{}, nil, nil, nil
+				return nil, nil, nil, nil
```
theorically a behaviour change- `nil` channel blocks whereas `Wait` on empty waitgroup succeeds immediately. `ClosedDoneChan`?

[Applies to a few places]

joamaki replied:
Went through all of these and changed them.

joamaki committed:
controller: Add done channel to UpdateController

To allow waiting for controller to be executed, add a channel
that is closed and swapped when the controller runs.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/channels/doc.go:
```diff
@@ -0,0 +1,7 @@
+// Copyright Authors of Cilium
+
+/*
+Package channels provides common utilities related to channels.
```
Little bit hesitant to introduce a whole new package just for this. If anyone knows a place where this would fit let me know. I suppose could just add this into `pkg/controller` instead as with the latest approach that's where `DoneChan` is used from... And call it `TriggeredChan` or something more descriptive.
